service/history/historyEventNotifier.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/pborman/uuid"
+	gen "github.com/uber/cadence/.gen/go/shared"
+)
+
+const (
+	eventsChanSize = 1000
+
+	// used for workflow pubsub status
+	statusIdle    int32 = 0
+	statusStarted int32 = 1
+	statusStopped int32 = 2
+)
+
+type (
+	workflowIdentifier struct {
+		domainID   string
+		workflowID string
+		runID      string
+	}
+
+	historyEvent struct {
+		workflowIdentifier
+		nextEventID       int64
+		isWorkflowRunning bool
+	}
+
+	historyEventNotifierImpl struct {
+		status int32
+		// stop signal channel
+		closeChan chan bool
+
+		sync.Mutex
+		// this channel will never close
+		eventsChan chan *historyEvent
```

Comment:
Since this is per host we can cleanly close the channel after all shards are shutdown.

Reply:
if i close the eventsChan, then any attempt to send notification to this channel will panic.

moreover, one daemon's shutdown depending on another daemon's shutdown may cause weird issue if anyone change the order of shutdown in history handler Stop() method